src/inc/Lfo.h

```diff
@@ -0,0 +1,88 @@
+#if !defined(__Lfo_hdr__)
+#define __Lfo_hdr__
+
+#define _USE_MATH_DEFINES
+#include <math.h>
+
+#include "ErrorDef.h"
+#include "Synthesis.h"
+#include "RingBuffer.h"
+
+class CLfo
+{
+public:
+    CLfo(float fModFreq, float fSampleRate)
+    {
+        m_pCRingBuffer = 0;
+        m_fRead = 0;
+        m_fModFreq = fModFreq;
+        m_fSampleRate = fSampleRate;
+        if (m_fModFreq == 0) {
+            m_iBufferLength = 1;
+        } else {
+            m_iBufferLength = static_cast<int>(1 / m_fModFreq);
+        }
+        m_pCRingBuffer = new CRingBuffer<float>(m_iBufferLength);
+    }
+    
+    ~CLfo()
+    {
+        delete m_pCRingBuffer;
+        m_pCRingBuffer = 0;
+    }
+    
+    Error_t setParam(float fParamValue)
+    {
+        m_fModFreq = fParamValue;
+        float *pSine = new float [m_iBufferLength];
+        
+        if (m_fModFreq == 0) {
+            m_iBufferLength = 1;
+        }
+        else {
+            m_iBufferLength = static_cast<int>(1 / m_fModFreq);
+        }
+        
+        CSynthesis::generateSine(pSine, m_fModFreq * m_fSampleRate, m_fSampleRate, m_iBufferLength);
+        m_pCRingBuffer-> putPostInc(pSine, m_iBufferLength);
+        delete [] pSine;
+        
+        return Error_t::kNoError;
+    }
```

Comment:
It looks like the only settable parameter here is the mod frequency. Changing the function definition to maybe `Error_t setParam(Lfo::LfoParam_t eParam, float fParamValue)` and then having the option to change parameters like the sample rate and amplitude would be more useful I think

Reply:
Since LFO is not a big project, I didn't apply the setParam function, instead, you can reset() then init() with your parameters then generate(). That way you should be able to set values.
```
Lfo.h - Line 33
    Error_t init (float fModFreqRatio, float fSampleRate) {
        this -> reset();
        if (fModFreqRatio <= 0 || fSampleRate <= 0) {
            return Error_t::kFunctionInvalidArgsError;
        }
        m_bIsInitialized = true;
        m_fModFreqRatio = fModFreqRatio;
        m_fSampleRate = fSampleRate;
        m_iWaveLength = static_cast<int>(1/fModFreqRatio);
        m_pSine = new float [m_iWaveLength];
        m_pRingBuffer = new CRingBuffer<float>(m_iWaveLength);
        return Error_t::kNoError;
    }
    
    Error_t reset () {
        if (m_bIsInitialized == true) {
            delete m_pRingBuffer;
            m_pRingBuffer = 0;
            delete [] m_pSine;
            m_pSine = 0;
        }
        m_fSampleRate = 0;
        m_fModFreqRatio = 0;
        m_bIsInitialized = false;
        return Error_t::kNoError;
    }
    
    Error_t generate () {
        CSynthesis::generateSine(m_pSine, m_fModFreqRatio * m_fSampleRate, m_fSampleRate, m_iWaveLength);
        m_pRingBuffer -> putPostInc(m_pSine, m_iWaveLength);
    }
```